magicbot/state_machine.py

```diff
@@ -514,21 +519,21 @@ def execute(self):
         new_state_start = tm
         
         # determine if the time has passed to execute the next state
-        # -> intentionally comes first, 
-        if state is not None and state.expires < tm:
-
-            previous_state = state
+        # -> intentionally comes first
+        if state is not None and state.ran and state.expires < tm:
+            new_state_start = state.expires
 
             if state.next_state is None:
-                state = None
+                # If the state expires and it's the last state, if the machine
+                # is still engaged then it should cycle back to the beginning
+                if self.__should_engage:
+                    self.next_state(self.__first)
+                    state = self.__state
+                else:
+                    state = None
             else:
                 self.next_state(state.next_state)
-                new_state_start = state.expires
                 state = self.__state
-
-            # Reset the expired time to prevent the state from expiring
-            # immediately if it's ran a second time
-            previous_state.expires = 0xffffffff
```

Comment:
So this did fix an issue we had where timed states wouldn't work correctly in a state machine that was executed more than once…

Reply:
See my comment on #88